tensorflow/stream_executor/cuda/cuda_dnn.cc

```diff
@@ -4464,19 +4519,32 @@ port::Status CreateOpRunners(
   };
 
   if (!use_fallback) {
-    auto heuristics = cudnn_frontend::EngineHeuristicsBuilder()
-                          .setOperationGraph(*op_graph)
-                          .setHeurMode(GetCudnnFrontendHeurMode())
-                          .build();
-    RETURN_MSG_IF_CUDNN_ERROR(heuristics);
+    std::optional<cudnn_frontend::EngineHeuristics> heuristics;
+    auto mode_b = cudnn_frontend::EngineHeuristicsBuilder()
+                      .setOperationGraph(*op_graph)
+                      .setHeurMode(GetCudnnFrontendHeurMode())
+                      .build();
+    if (mode_b.get_status() != CUDNN_STATUS_SUCCESS) {
+      // The mode_b can fallback to mode_instant if it cannot support things.
+      // However, there is a known issue in cudnn when dealing with runtime
+      // fusion engines. So, we manually do that here.
+      // TODO(kaixih@nvidia): remove this when the cudnn fixes it.
+      heuristics.emplace(cudnn_frontend::EngineHeuristicsBuilder()
+                             .setOperationGraph(*op_graph)
+                             .setHeurMode(CUDNN_HEUR_MODE_INSTANT)
+                             .build());
+      RETURN_MSG_IF_CUDNN_ERROR(*heuristics);
+    } else {
+      heuristics.emplace(std::move(mode_b));
+    }
```

Comment:
After this block, it's guaranteed that `heuristics.has_value() == true`.  Therefore, heuristics isn't really an optional.

Suggest writing this as an IIFE.

```
cudnn_frontend::EngineHeuristics heuristics = [&] {
  auto heuristics = cudnn_frontend::EngineHeuristicsBuilder()
                      .setOperationGraph(*op_graph)
                      .setHeurMode(GetCudnnFrontendHeurMode())
                      .build();
  if (heuristics.status() == CUDNN_STATUS_SUCCESS) { heuristics; }
  return cudnn_frontend::EngineHeuristicsBuilder()
                             .setOperationGraph(*op_graph)
                             .setHeurMode(CUDNN_HEUR_MODE_INSTANT)
                             .build();
}();
```

This new logic makes it much clearer what's going on.  It also guides us to writing a better comment.  I think what they mean to say is something like the following.

> In theory, mode GetCudnnFrontendHeurMode() is supposed to fall back to HEUR_MODE_INSTANT if it can't support things [what things??].  But there's a known issue where it doesn't when dealing with runtime fusion engines [what is that?].  So we do it manually here.